File: src/TradingSystem.Web/Services/ThemeService.cs
using Microsoft.JSInterop;

namespace TradingSystem.Web.Services;

public interface IThemeService
{
    event Action? OnThemeChanged;
    string CurrentTheme { get; }
    bool IsDarkMode { get; }
    Task InitializeAsync();
    Task ToggleThemeAsync();
    Task SetThemeAsync(string theme);
}

public class ThemeService : IThemeService
{
    private readonly IJSRuntime _jsRuntime;
    private const string StorageKey = "trading_theme";
    private string _currentTheme = "dark";
    private bool _isInitialized = false;

    public event Action? OnThemeChanged;
    public string CurrentTheme => _currentTheme;
    public bool IsDarkMode => _currentTheme == "dark";

    public ThemeService(IJSRuntime jsRuntime)
    {
        _jsRuntime = jsRuntime;
    }

    public async Task InitializeAsync()
    {
        if (_isInitialized) return;

        try
        {
            var savedTheme = await _jsRuntime.InvokeAsync<string?>("localStorage.getItem", StorageKey);
            _currentTheme = savedTheme ?? "dark";
            await ApplyThemeAsync();
        }
        catch
        {
            _currentTheme = "dark";
        }
        _isInitialized = true;
    }

    public async Task ToggleThemeAsync()
    {
        _currentTheme = _currentTheme == "dark" ? "light" : "dark";
        await SaveAndApplyAsync();
    }

    public async Task SetThemeAsync(string theme)
    {
        if (theme != "dark" && theme != "light") return;
        _currentTheme = theme;
        await SaveAndApplyAsync();
    }

    private async Task SaveAndApplyAsync()
    {
        await _jsRuntime.InvokeVoidAsync("localStorage.setItem", StorageKey, _currentTheme);
        await ApplyThemeAsync();
        OnThemeChanged?.Invoke();
    }

    private async Task ApplyThemeAsync()
    {
        await _jsRuntime.InvokeVoidAsync("eval", $"document.documentElement.setAttribute('data-theme', '{_currentTheme}')");
    }
}

File: src/TradingSystem.Angular/src/app/services/theme.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';

@Injectable({ providedIn: 'root' })
export class ThemeService {
  private readonly STORAGE_KEY = 'trading_theme';
  private themeSubject = new BehaviorSubject<string>(this.getSavedTheme());

  theme$ = this.themeSubject.asObservable();

  get currentTheme(): string { return this.themeSubject.value; }
  get isDarkMode(): boolean { return this.themeSubject.value === 'dark'; }

  initialize(): void {
    this.applyTheme(this.themeSubject.value);
  }

  toggleTheme(): void {
    const newTheme = this.isDarkMode ? 'light' : 'dark';
    this.setTheme(newTheme);
  }

  setTheme(theme: string): void {
    if (theme !== 'dark' && theme !== 'light') return;
    localStorage.setItem(this.STORAGE_KEY, theme);
    this.themeSubject.next(theme);
    this.applyTheme(theme);
  }

  private applyTheme(theme: string): void {
    document.documentElement.setAttribute('data-theme', theme);
  }

  private getSavedTheme(): string {
    return localStorage.getItem(this.STORAGE_KEY) || 'dark';
  }
}
